packages/medusa/src/services/csv-parser.ts

```diff
@@ -0,0 +1,161 @@
+import { AwilixContainer } from "awilix"
+import { difference } from "lodash"
+import Papa, { ParseConfig } from "papaparse"
+import { AbstractParser } from "../interfaces/abstract-parser"
+import { CsvSchema } from "../interfaces/csv-parser"
+
+const DEFAULT_PARSE_OPTIONS = {
+  dynamicTyping: true,
+  header: true,
+}
+
+class CsvParser<
+  TSchema extends CsvSchema = CsvSchema,
+  TParserResult = unknown,
+  TOutputResult = unknown
+> extends AbstractParser<TSchema, TParserResult, ParseConfig, TOutputResult> {
+  protected readonly $$delimiter: string = ";"
+
+  constructor(
+    protected readonly container: AwilixContainer,
+    schema: TSchema,
+    delimiter?: string
+  ) {
+    super(schema)
+    if (delimiter) {
+      this.$$delimiter = delimiter
+    }
+  }
+
+  public async parse(
+    readableStream: NodeJS.ReadableStream,
+    options: ParseConfig = DEFAULT_PARSE_OPTIONS
+  ): Promise<TParserResult[]> {
+    const csvStream = Papa.parse(Papa.NODE_STREAM_INPUT, options)
+
+    const parsedContent: TParserResult[] = []
+    readableStream.pipe(csvStream)
+    for await (const chunk of csvStream) {
+      parsedContent.push(chunk)
+    }
+
+    return parsedContent
+  }
+
+  async buildData(data: TParserResult[]): Promise<TOutputResult[]> {
+    const validatedData = [] as TOutputResult[]
+    for (let i = 0; i < data.length; i++) {
+      const builtLine = await this._buildLine(data[i], i + 1)
+      validatedData.push(builtLine)
+    }
+    return validatedData
+  }
+
+  private async _buildLine(
+    line: TParserResult,
+    lineNumber: number
+  ): Promise<TOutputResult> {
+    const outputTuple = {} as TOutputResult
+    const columnMap = this._buildColumnMap(this.$$schema.columns)
+
+    const tupleKeys = Object.keys(line)
+
+    /**
+     * map which keeps track of the columns processed
+     * used to detect any missing columns which are present in the schema but not in the line
+     */
+    const processedColumns = {}
+    for (const tupleKey of tupleKeys) {
+      const column = this._resolveColumn(tupleKey, columnMap)
+
+      /**
+       * if the tupleKey does not correspond to any column defined in the schema
+       */
+      if (!column) {
+        throw new Error(
+          `Unable to treat column ${tupleKey} from the csv file. No target column found in the provided schema`
+        )
+      }
+
+      /**
+       * if the value corresponding to the tupleKey is empty and the column is required in the schema
+       */
+      if (!line[tupleKey] && column.required) {
+        throw new Error(
+          `No value found for target column "${column.name}" in line ${lineNumber} of the given csv file`
+        )
+      }
+
+      const context = {
+        line,
+        lineNumber,
+        column: tupleKey,
+      }
+
+      if (column.validator) {
+        await column.validator.validate(line[tupleKey], context)
+      }
+
+      outputTuple[column.mapTo ?? tupleKey] = column.transformer
```

Comment:
@fPolic If I understood the question, I believe we'll run into an issue if the column definition in the schema also has a `mapTo` property set, then it that case, you are right. Multiple column values in a tuple will all be added to the same object property (which comes from the `mapTo` property). 

In case you meant something else, can you please check this example test case and tell me if it captures what you had in mind?
https://github.com/medusajs/medusa/blob/92404e4a1d0f210fa9ae8b9aeb171d8b5be6a67d/packages/medusa/src/services/__tests__/csv-parser.js#L184-L215

Reply:
I think that what @fPolic means is that we could add `isCollection: boolean` and `findIndex` to the schema so that we can configure that the value must be part of an object that is located to the `findIndex` result in the collection. That would give something around the line of:

```ts
    if (column.isCollection) {
      outputTuple[column.name] = outputTuple[name] || []
      const targetIndex = column.findIndex(outputTuple[column.name], tupleKey)
      if (targetIndex === -1) {
        throw new Error(`Unable to find the object on which assigning the property ${column.mapTo ?? tupleKey}`)
      }
    
      outputTuple[column.name][targetIndex][column.mapTo ?? tupleKey] = column.transformer
        ? column.transformer(line[tupleKey], context)
        : line[tupleKey]
    } else {
      outputTuple[column.mapTo ?? tupleKey] = column.transformer
        ? column.transformer(line[tupleKey], context)
        : line[tupleKey]
    }
```

wdyt guys?

Reply:
@zakariaelas yes this is the exact case - when the column definition also has `mapTo` prop. If a column definition has `mapTo` and `match` prop all values from columns that matched should be added, wdyt?

@adrien2p should we assume that it is a collection if we have `mapTo` + `match`

Reply:
I think that it should be explicit to avoid forgetting about it, plus we need a findIndex method to return an object from a collection ok which to attach the property

Reply:
The thing we implicit features is that as a user, you will have to look at the impl to understand how to manage a collection, also, if the user is lazy and just use the title as a description plus the column name, he will probably use the match all the time 😂

Reply:
Interesting. Do you guys believe we should make this "aggregation" of different columns into an array as part of the `CsvParser`? One other option would be to simply leave this "aggregation" to the consumer and instead aim for solving the `mapTo` issue a bit differently. For example, one thing we could do is to make the `mapTo` property also take a function, and call it when building data. To explain my thoughts in code, take a look a the snippet below:

```typescript
export type CsvSchemaColumn<TLine> = {
  // ...
  mapTo?: (tupleKey: string, context: CsvParserContext<TLine>) => string
  // ...
}


const csvParser = new CsvParser({}, { 
   columns: [ 
     { 
       name: "title", 
     }, 
     { 
       name: "variants", 
       match: /.*Variant Price.*/i, 
       transformer: (value) => Math.round(Number(value) * 100), 
       
       // now the initial issue is fixed because the different tuple keys that match this column will not map to the same column
       mapTo: (tupleKey: string, context: CsvParserContext<TLine>) => {
          // e.g: "Variant Price USD" -> variant_price_usd
          return tupleKey.toLoweCase().join('_')
       }
     }, 
   ], 
 }) 
```

So just we get aligned on the desired API, do you guys believe we should support:

1. Aggregating all the tuple columns that match a schema column into an array/list using an option like `isCollection: boolean` or `aggregate: boolean` 
2. Extend the `mapTo` property to include functions in order to support "dynamic" mapping of columns (just like in the code snippet above). Doing this would assume that these aggregations fall within the responsibility of the consumers and not the `CsvParser` (which is reasonable).
3. Both 1. and 2.

Personally I'm totally fine with either, but I slightly lean towards Option 2 or at least 3. I see Option 1. as a nice to have, but not necessarily as something that solves the root problem. It feels as if we're saying: _"as a consumer, you can't use match and mapTo WITHOUT aggregating"_.

What do you guys think?

Reply:
I think that the parser should allow to do that through the schema. Otherwise, you can end up having to re do the schema to reorganise the output. 
your idea just gave me an idea, what if we remove the mapTo and add assign which would be a function that take an input object (empty the first time) and the validated value? We can then also remove the transformer.

this approach would give full control to the consumer of what the output object should looks like. 

I think it simplify everything, allow more customisation from the consumer and allow to scale and evolve pretty easily without having to always update the impl of the parser.


Wdyt guys?

Reply:
As a consumer of this service, I was hoping that the parser would return structured data without the need to match/reshape/aggregate data again. That being said, I also like the general idea regarding this function that "hooks" into the parsing process and allows for the IoC 😄.